wrappers/services.go

```diff
@@ -705,41 +706,160 @@ func makeAbbrevWeekdays(start time.Weekday, end time.Weekday) []string {
 	return out
 }
 
+// daysRange generates a string representing a continuous range between given
+// day numbers, which due to compatiblilty with old systemd version uses a
+// verbose syntax of x,y,z instead of x..z
+func daysRange(start, end uint) string {
+	var buf bytes.Buffer
+	for i := start; i <= end; i++ {
+		buf.WriteString(strconv.FormatInt(int64(i), 10))
+		if i < end {
+			buf.WriteRune(',')
+		}
+	}
+	return buf.String()
+}
+
 // generateOnCalendarSchedules converts a schedule into OnCalendar schedules
 // suitable for use in systemd *.timer units using systemd.time(7)
 // https://www.freedesktop.org/software/systemd/man/systemd.time.html
+// XXX: old systemd versions do not support x..y ranges
 func generateOnCalendarSchedules(schedule []*timeutil.Schedule) []string {
 	calendarEvents := make([]string, 0, len(schedule))
 	for _, sched := range schedule {
 		days := make([]string, 0, len(sched.WeekSpans))
 		for _, week := range sched.WeekSpans {
 			abbrev := strings.Join(makeAbbrevWeekdays(week.Start.Weekday, week.End.Weekday), ",")
-			switch week.Start.Pos {
-			case timeutil.EveryWeek:
+
+			if week.Start.Pos == timeutil.EveryWeek && week.End.Pos == timeutil.EveryWeek {
 				// eg: mon, mon-fri, fri-mon
 				days = append(days, fmt.Sprintf("%s *-*-*", abbrev))
-			case timeutil.LastWeek:
-				// eg: mon5
-				days = append(days, fmt.Sprintf("%s *-*~7/1", abbrev))
-			default:
-				// eg: mon1, fri1, mon1-tue2
-				startDay := (week.Start.Pos-1)*7 + 1
-				endDay := week.End.Pos * 7
-
-				// NOTE: schedule mon1-tue2 (all weekdays
-				// between the first Monday of the month, until
-				// the second Tuesday of the month) is not
-				// translatable to systemd.time(7) format, for
-				// this assume all weekdays and allow the runner
-				// to do the filtering
-				if week.Start != week.End {
+				continue
+			}
+			// examples:
+			// mon1 - Mon *-*-1..7 (Monday during the first 7 days)
+			// fri1 - Fri *-*-1..7 (Friday during the first 7 days)
+
+			// entries below are will make systemd timer expire more
+			// frequently than the schedule suggests, however snap
+			// runner evaluates current time and gates the actual
+			// action
+			//
+			// mon1-tue - *-*-1..7 *-*-8 (anchored at first
+			// Monday; Monday happens during the 7 days,
+			// Tuesday can possibly happen on the 8th day if
+			// the month started on Tuesday)
+			//
+			// mon-tue1 - *-*~1 *-*-1..7 (anchored at first
+			// Tuesday; matching Monday can happen on the
+			// last day of previous month if Tuesday is the
+			// 1st)
+			//
+			// mon5-tue - *-*~1..7 *-*-1 (anchored at last
+			// Monday, the matching Tuesday can still happen
+			// within the last 7 days, or on the 1st of the
+			// next month)
+			//
+			// fri4-mon - *-*-22-31 *-*-1..7 (anchored at 4th
+			// Friday, can span onto the next month, extreme case in
+			// February when 28th is Friday)
+			//
+			// XXX: since old versions of systemd, eg. 229 available
```

Comment:
Can we detect systemd version and only apply workaround for the old ones?

Reply:
I think that'd be too fragile. Systemd is maintained outside of snapd on a classic system, so there's a possibility that the version may go back to one that does not support the improved syntax.